test/compilable/b12647.d

```diff
@@ -0,0 +1,9 @@
+void assertThrown12647(E)(lazy E expression)
+{
+    expression();
+}
+
+void test12647() nothrow pure @safe @nogc
+{
+    assertThrown12647(0);
+}
```

Comment:
You should also add a test for an expression that throws, and one that allocates (separately) to ensure that it doesn't accept-invalid.